### PR TITLE
[FEATURE ds-extended-errors] Make adapter error extendable and add more error types

### PR DESCRIPTION
**Pull request by tchak**
This is going to help with questions like #3493 or #3491

### CODE REVIEW

**Comment by btecu** [comment]
This could probably be labeled as [BUGFIX release].

**Comment by tchak** [comment]
Strictly this is not a bug fix. I am not sure where is the line of what we back porting in 1.13 series. @bmac  you have an opinion?

**Comment by bmac** [comment]
I think we should treat it as a new feature and feature flag it for a future 2.x release. If we find after it is release that it would provide value and help with the transition for users on the 1.13 series we can always backport it in the future.

**Comment by bmac** [comment]
@tchak What is the status of this pr?

**Comment by lindyhopchris** [comment]
@tchak  @bmac can I also ask what the status of this is? This is really required because `AdapterError` is just too generic to handle effectively in route transitions. Plus because it doesn't give access to the HTTP status code, it's difficult to handle.

For info, I do think that `AdapterError` should have the original HTTP status code on it. The reason being that the JSON API spec says:

> Error objects provide additional information about problems encountered while performing an operation. 

The key word here being **additional**. There's no requirement for a server to send a response that does include error objects within the JSON body - a simple HTTP status response with no content is valid. I know you've said in #3491 that the original HTTP response should not leak into the application, but it seems to me that the status code is required information considering that there is no obligation under the spec for the server to return error objects in its response body.

**Comment by bmac** [comment]
@tchak  Do you mind rebasing this pr and wrapping it in a [feature flag](https://github.com/emberjs/data/pull/3979)?

**Comment by tchak** [comment]
@bmac done

File: addon/adapters/rest.js [hunk 1]
```diff
@@ -906,7 +910,18 @@ var RESTAdapter = Adapter.extend(BuildURLMixin, {
     let errors          = this.normalizeErrorResponse(status, headers, payload);
     let detailedMessage = this.generatedDetailedMessage(status, headers, payload, requestData);
 
-    return new AdapterError(errors, detailedMessage);
+    switch (status) {
```
Comment (bmac) [inline addon/adapters/rest.js]: Should this be wrapped in a `isEnabled('ds-extended-errors');`?

**Comment by lindyhopchris** [comment]
@tchak @bmac 

are there any plans to get this out from behind a feature flag? if so, what release might that happen in?

**Comment by bmac** [comment]
Currently the only think blocking this feature flag from being enabled is a lack of documentation. We need some more fleshed out API docs for all of the errors and a pr on the guides repo explaining the new errors.

I haven't had any time to add docs for this feature but @lindyhopchris if you are up for the challenge a pr updating the error's [doc blocks](https://github.com/emberjs/data/blob/c9d8212c857ca78218ad98d11621819b38dba98f/addon/adapters/errors.js#L119-L164) to describe when they trigger (and their [status codes](https://github.com/emberjs/data/blob/master/FEATURES.md#feature-flags)) would be a great place to start.

**Comment by wecc** [comment]
Related: #3585

**Comment by lindyhopchris** [comment]
@bmac happy to give it a go, will probably have some time later this week to look at it and submit a PR.

**Comment by sandstrom** [comment]
@bmac Now that it's documented (https://github.com/emberjs/data/issues/4632), can the flag be removed? 😄

**Comment by sandstrom** [comment]
friendly ping @alexlafroscia 😄

**Comment by lindyhopchris** [comment]
@sandstrom 

re: your second point, `AbortError` and `TimeoutError` are included:
https://github.com/emberjs/data/blob/master/addon/adapters/errors.js#L214-L226

and are used here:
https://github.com/emberjs/data/blob/master/addon/adapters/rest.js#L1518-L1520

**Comment by sandstrom** [comment]
@lindyhopchris Thanks, you are correct! Thanks for correcting me + this is great news! 😄

**Comment by lindyhopchris** [comment]
ping @bmac 

would be great to see this feature taken out from behind the feature flag. any update on when this might happen?

**Pull request merged**